common-ui-compose/src/main/java/app/tivi/common/compose/insets.kt

```diff
@@ -14,64 +14,410 @@
 @Composable
-fun ProvideInsets(
-    liveData: LiveData<WindowInsetsCompat?>,
-    children: @Composable () -> Unit
+fun ProvideDisplayInsets(
+    consumeInsets: Boolean = true,
+    content: @Composable () -> Unit
 ) {
-    val currentInsets = remember(liveData) {
-        liveData.map { if (it != null) InsetsHolder(it) else InsetsHolder() }
-            .distinctUntilChanged()
-    }.observeAsState(InsetsHolder())
+    val view = ViewAmbient.current
+
+    val displayInsets = remember { DisplayInsets() }
+
+    onCommit(view) {
+        ViewCompat.setOnApplyWindowInsetsListener(view) { _, windowInsets ->
+            displayInsets.systemBars.updateFrom(windowInsets.systemWindowInsets)
+            displayInsets.systemGestures.updateFrom(windowInsets.systemGestureInsets)
+
+            if (consumeInsets) {
+                // TODO: Move this to WindowInsetsCompat.CONSUMED
+                windowInsets.consumeSystemWindowInsets()
+            } else {
+                // Return the unconsumed insets
+                windowInsets
+            }
+        }
+
+        // Add an OnAttachStateChangeListener to request an inset pass each time we're attached
+        // to the window
+        val attachListener = object : View.OnAttachStateChangeListener {
+            override fun onViewAttachedToWindow(v: View) = v.requestApplyInsets()
+            override fun onViewDetachedFromWindow(v: View) = Unit
+        }
+        view.addOnAttachStateChangeListener(attachListener)
+
+        if (view.isAttachedToWindow) {
+            // If the view is already attached, we can request an inset pass now
+            view.requestApplyInsets()
+        }
+
+        onDispose {
+            // Once we leave the composition, remove our OnAttachStateChangeListener
+            view.removeOnAttachStateChangeListener(attachListener)
+        }
+    }
+
+    Providers(DisplayInsetsAmbient provides displayInsets) {
+        content()
+    }
+}
+
+/**
+ * Selectively apply additional space which matches the width/height of any system bars present
+ * on the respective edges of the screen.
+ *
+ * @param enabled Whether to apply padding using the system bar dimensions on the respective edges.
+ * Defaults to `true`.
+ */
+fun Modifier.systemBarsPadding(enabled: Boolean = true) = composed {
+    insetsPadding(
+        insets = DisplayInsetsAmbient.current.systemBars,
+        left = enabled,
+        top = enabled,
+        right = enabled,
+        bottom = enabled
+    )
+}
+
+/**
+ * Apply additional space which matches the height of the status height along the top edge
+ * of the content.
+ */
+fun Modifier.statusBarPadding() = composed {
+    insetsPadding(insets = DisplayInsetsAmbient.current.systemBars, top = true)
+}
+
+/**
+ * Apply additional space which matches the height of the navigation bar height
+ * along the [bottom] edge of the content, and additional space which matches the width of
+ * the navigation bar on the respective [left] and [right] edges.
```

Comment:
“width of the nav bar” sounds misleading as it's the overlap with insets right?

Reply:
Updated.